src/terminalOutput.ts

```diff
@@ -48,7 +53,18 @@ export class TerminalOutput {
         const browser = await ${browserName}.launch(${formatObjectOrVoid(launchOptions)});
         const context = await browser.newContext(${formatContextOptions(contextOptions, deviceName)});
       })();`);
-    this._out.write(this._highlight(formatter.format()) + '\n');
+    this._write(formatter.format(), '\n');
+  }
+
+  writeOutputFile() {
+    if (this._outputFile)
+      fs.writeFileSync(this._outputFile, this._toBeWrittenLines.join('\n'))
+  }
+
+  private _write(text: string, suffix: string = "") {
+    if (this._outputFile)
+      this._toBeWrittenLines.push(...(text + suffix).trimEnd().split('\n'))
+    this._out.write(this._highlight(text) + suffix)
```

Comment:
Let's throw a bit more design here: have 'fileOutput' and 'terminalOutput' generator 'Observers'. Observer will have { addLine:, popLine:, flush: } interface. Highlight will be implementation detail of the terminalOutput, with the highlight library only imported from there.

Reply:
good catch.